File: game/src/items/__tests__/items.test.ts
import { describe, expect, test } from 'vitest'
import { BLOCK } from '../../world/palette'
import { breakTime, dropForBlock, getItem, ITEMS } from '../registry'
import { Inventory, HOTBAR_SIZE, INVENTORY_SIZE } from '../inventory'

describe('registry', () => {
  test('every placeable block is an item that places itself', () => {
    expect(getItem('grass').block).toBe(BLOCK.grass)
    expect(getItem('ore_iron').block).toBe(BLOCK.ore_iron)
    expect(ITEMS.water).toBeUndefined()
    expect(ITEMS.air).toBeUndefined()
    expect(() => getItem('nope')).toThrow()
  })

  test('block drops: grass→dirt, ores→resources, water→nothing, others→themselves', () => {
    expect(dropForBlock(BLOCK.grass)).toBe('dirt')
    expect(dropForBlock(BLOCK.ore_coal)).toBe('coal')
    expect(dropForBlock(BLOCK.ore_iron)).toBe('iron')
    expect(dropForBlock(BLOCK.water)).toBeNull()
    expect(dropForBlock(BLOCK.stone)).toBe('stone')
    expect(dropForBlock(BLOCK.leaves)).toBe('leaves')
  })

  test('break times: pickaxe tiers speed digging; stone needs a pickaxe', () => {
    expect(breakTime(BLOCK.dirt, null)).toBeCloseTo(0.75)
    expect(breakTime(BLOCK.dirt, 'pickaxe_wood')).toBeCloseTo(0.5)
    expect(breakTime(BLOCK.stone, null)).toBe(Infinity)
    expect(breakTime(BLOCK.stone, 'sword')).toBe(Infinity)
    expect(breakTime(BLOCK.stone, 'pickaxe_wood')).toBeCloseTo(5 / 1.5)
    expect(breakTime(BLOCK.stone, 'pickaxe_iron')).toBeCloseTo(1.25)
    expect(breakTime(BLOCK.water, 'pickaxe_iron')).toBe(Infinity)
  })
})

describe('inventory', () => {
  test('stacks fill up to maxStack before opening a new slot', () => {
    const inv = new Inventory()
    expect(inv.add('dirt', 70)).toBe(0)
    expect(inv.get(0)).toEqual({ id: 'dirt', count: 64 })
    expect(inv.get(1)).toEqual({ id: 'dirt', count: 6 })
    expect(inv.count('dirt')).toBe(70)
    expect(inv.hotbar()).toHaveLength(HOTBAR_SIZE)
    expect(inv.all()).toHaveLength(INVENTORY_SIZE)
  })

  test('tools do not stack and a full inventory reports leftovers', () => {
    const inv = new Inventory()
    inv.add('sword', 1)
    inv.add('sword', 1)
    expect(inv.get(0)).toEqual({ id: 'sword', count: 1 })
    expect(inv.get(1)).toEqual({ id: 'sword', count: 1 })
    for (let i = 0; i < INVENTORY_SIZE - 2; i++) inv.add('rifle', 1)
    expect(inv.add('dirt', 5)).toBe(5)
  })

  test('takeFromSlot and remove update counts and clear empty slots', () => {
    const inv = new Inventory()
    inv.add('planks', 10)
    expect(inv.takeFromSlot(0, 3)).toBe(3)
    expect(inv.get(0)).toEqual({ id: 'planks', count: 7 })
    expect(inv.takeFromSlot(0, 99)).toBe(7)
    expect(inv.get(0)).toBeNull()
    expect(inv.takeFromSlot(0, 1)).toBe(0)

    inv.add('ammo', 100) // 100 in one slot (max 120)
    inv.add('ammo', 50)  // 20 top up + 30 new slot
    expect(inv.remove('ammo', 200)).toBe(false)
    expect(inv.count('ammo')).toBe(150)
    expect(inv.remove('ammo', 40)).toBe(true)
    expect(inv.count('ammo')).toBe(110)
  })

  test('version bumps only on real changes', () => {
    const inv = new Inventory()
    const v0 = inv.version
    inv.remove('dirt', 1)
    expect(inv.version).toBe(v0)
    inv.add('dirt', 1)
    expect(inv.version).toBe(v0 + 1)
    inv.swap(0, 5)
    expect(inv.get(5)).toEqual({ id: 'dirt', count: 1 })
    expect(inv.get(0)).toBeNull()
  })
})

File: client/test/items/items_test.dart
import 'package:block_survival/items/inventory.dart';
import 'package:block_survival/items/recipes.dart';
import 'package:block_survival/items/registry.dart';
import 'package:block_survival/world/ores.dart';
import 'package:block_survival/world/palette.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('block items exist for every placeable block and nothing else', () {
    expect(getItem('grass').block, Block.grass);
    expect(getItem('planks').kind, ItemKind.block);
    expect(items.containsKey('water'), isFalse);
    expect(items.containsKey('bedrock'), isFalse);
    expect(getItem('torch').kind, ItemKind.prop);
    expect(getItem('torch').block, Block.torch);
    expect(items.length, 44); // 20 placeable blocks + 24 others, as on the web
  });

  test('drops and break times follow the web rules', () {
    expect(dropForBlock(Block.grass), 'dirt');
    expect(dropForBlock(Block.stone), 'cobble');
    expect(dropForBlock(Block.oreIron), 'iron');
    expect(dropForBlock(Block.water), isNull);
    expect(breakTime(Block.stone, null), double.infinity);
    expect(breakTime(Block.stone, 'pickaxe_wood'), closeTo(5 / 1.5, 1e-9));
    expect(breakTime(Block.dirt, null), 0.75);
    expect(breakTime(Block.bedrock, 'pickaxe_iron'), double.infinity);
  });

  // issue #25: the same ladder the web client's minerals test pins
  test('every mineral drops, and the deep ones are gated behind iron', () {
    for (final o in ores) {
      expect(dropForBlock(o.block), o.drop, reason: '${o.drop} drop');
      expect(items.containsKey(o.drop), isTrue, reason: '${o.drop} item');
      expect(breakTime(o.block, null), double.infinity);
    }
    expect(breakTime(Block.oreCoal, 'pickaxe_wood'), lessThan(double.infinity));
    expect(breakTime(Block.oreCopper, 'pickaxe_wood'), double.infinity);
    expect(
      breakTime(Block.oreCopper, 'pickaxe_stone'),
      lessThan(double.infinity),
    );
    for (final ore in [
      Block.oreGold,
      Block.oreRedstone,
      Block.oreDiamond,
      Block.oreEmerald,
    ]) {
      expect(breakTime(ore, 'pickaxe_copper'), double.infinity);
      expect(breakTime(ore, 'pickaxe_iron'), lessThan(double.infinity));
    }
    // gold digs faster than iron but cannot touch what iron opens
    expect(
      breakTime(Block.stone, 'pickaxe_gold'),
      lessThan(breakTime(Block.stone, 'pickaxe_iron')),
    );
    expect(breakTime(Block.oreDiamond, 'pickaxe_gold'), double.infinity);
    expect(mineTierOf('pickaxe_gold'), lessThan(mineTierOf('pickaxe_iron')));
  });

  test('a prospector senses, and the attuned one senses further', () {
    expect(getItem('prospector').senseRange, greaterThan(0));
    expect(
      getItem('prospector_far').senseRange!,
      greaterThan(getItem('prospector').senseRange!),
    );
  });

  test('every mineral has a recipe to go into', () {
    final used = {for (final r in recipes) ...r.inputs.map((i) => i.id)};
    for (final o in ores) {
      expect(used.contains(o.drop), isTrue, reason: '${o.drop} is unused');
    }
  });

  test('inventory stacks, overflows and removes like the web one', () {
    final inv = Inventory();
    var changes = 0;
    inv.addListener(() => changes++);
    expect(inv.add('log', 70), 0);
    expect(inv.get(0), const ItemStack('log', 64));
    expect(inv.get(1), const ItemStack('log', 6));
    expect(inv.count('log'), 70);
    expect(inv.add('sword', 1), 0);
    expect(inv.get(2), const ItemStack('sword', 1));
    expect(inv.remove('log', 10), isTrue);
    expect(inv.get(1), isNull, reason: 'removed from the last slot first');
    expect(inv.get(0), const ItemStack('log', 60));
    expect(inv.remove('log', 61), isFalse);
    expect(inv.takeFromSlot(2, 5), 1);
    inv.swap(0, 8);
    expect(inv.get(8), const ItemStack('log', 60));
    expect(inv.hotbar().length, hotbarSize);
    expect(changes, greaterThan(4));
  });

  test('crafting: hand vs bench, status and consumption', () {
    final inv = Inventory();
    expect(recipesFor(CraftContext.hand).map((r) => r.id), [
      'planks',
      'stick',
      'workbench',
      'pickaxe_wood',
    ]);
    expect(
      recipesFor(CraftContext.bench).any((r) => r.id == 'workbench'),
      isFalse,
    );
    final planks = getRecipe('planks');
    expect(craftStatus(inv, planks, nearBench: false), CraftStatus.missing);
    inv.add('log', 1);
    expect(craftStatus(inv, planks, nearBench: false), CraftStatus.ok);
    expect(craft(inv, planks, nearBench: false), 0);
    expect(inv.count('log'), 0);
    expect(inv.count('planks'), 4);
    expect(
      craftStatus(inv, getRecipe('torch'), nearBench: false),
      CraftStatus.needsBench,
    );
    expect(craft(inv, getRecipe('torch'), nearBench: false), isNull);
  });
}
